File: src/converter/plugins/convertHeritageClause.ts
import ts from "typescript";
import {createSimplePlugin} from "../plugin.js";
import {CheckCoverageService, checkCoverageServiceKey} from "./CheckCoveragePlugin.js";

export const convertHeritageClause = createSimplePlugin((node, context, render) => {
    if (!ts.isHeritageClause(node)) return null

    const checkCoverageService = context.lookupService<CheckCoverageService>(checkCoverageServiceKey)
    checkCoverageService?.cover(node)

    return node.types
        .map(type => render(type))
        .filter(Boolean)
        .join(", ")
})

File: karakum/src/jsMain/kotlin/io/github/sgrishchenko/karakum/extension/plugins/convertHeritageClause.kt
package io.github.sgrishchenko.karakum.extension.plugins

import io.github.sgrishchenko.karakum.extension.createPlugin
import typescript.asArray
import typescript.isHeritageClause

val convertHeritageClause = createPlugin plugin@{ node, context, render ->
    if (!isHeritageClause(node)) return@plugin null

    val checkCoverageService = context.lookupService(checkCoverageServiceKey)
    checkCoverageService?.cover(node)

    node.types.asArray()
        .map { render(it) }
        .filter { it.isNotEmpty() }
        .joinToString(separator = ", ")
}
